File: apps/player-android/app/src/test/java/org/tilecast/player/content/CountdownBarTargetTimeTest.kt
package org.tilecast.player.content

import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Test
import org.tilecast.player.network.ManifestPlugin
import org.tilecast.player.network.ManifestPluginConfig
import java.time.Instant

class CountdownBarTargetTimeTest {
    private fun weekly(targetTime: String) =
        ManifestPlugin(
            id = "bar-seconds",
            type = "countdown_bar",
            version = 1,
            config =
                ManifestPluginConfig(
                    name = "Lunch",
                    message = "Lunch ends in",
                    scheduleType = "weekly",
                    targetTime = targetTime,
                    daysOfWeek = listOf(1),
                    timezone = "America/New_York",
                    leadTimeSeconds = 900,
                    completionText = "Lunch is over",
                    displayMode = "overlay",
                    heightPx = 72,
                    priority = 10,
                ),
        )

    @Test
    fun `preserves seconds from PostgreSQL-style target times`() {
        val active =
            resolveCountdownBar(
                listOf(weekly("12:00:45")),
                Instant.parse("2026-07-27T15:50:00Z"),
            )

        assertEquals(Instant.parse("2026-07-27T16:00:45Z"), active?.targetAt)
        assertEquals("10m 45s", active?.value)
    }

    @Test
    fun `rejects trailing text instead of treating it as a valid time`() {
        assertNull(
            resolveCountdownBar(
                listOf(weekly("12:00oops")),
                Instant.parse("2026-07-27T15:50:00Z"),
            ),
        )
    }
}

File: apps/player-linux/src/renderer/countdown-bar-target-time.test.ts
import { describe, expect, it } from "vitest";
import "./countdown-display";
import "./countdown-bar-resolver";

interface CountdownBarPlugin {
  id: string;
  type: string;
  version: number;
  config: {
    message: string;
    scheduleType: "weekly";
    targetTime: string;
    daysOfWeek: number[];
    timezone: string;
    leadTimeSeconds: number;
    completionText: string;
    displayMode: "overlay";
    heightPx: number;
    priority: number;
  };
}

interface CountdownBarResolver {
  resolve(
    plugins: CountdownBarPlugin[],
    now: Date,
  ): { targetAt: string; value: string } | null;
}

const resolver = (
  globalThis as typeof globalThis & {
    tilecastCountdownBar: CountdownBarResolver;
  }
).tilecastCountdownBar;

function weekly(targetTime: string): CountdownBarPlugin {
  return {
    id: "bar-seconds",
    type: "countdown_bar",
    version: 1,
    config: {
      message: "Lunch ends in",
      scheduleType: "weekly",
      targetTime,
      daysOfWeek: [1],
      timezone: "America/New_York",
      leadTimeSeconds: 900,
      completionText: "Lunch is over",
      displayMode: "overlay",
      heightPx: 72,
      priority: 10,
    },
  };
}

describe("countdown bar target time parsing", () => {
  it("preserves seconds from PostgreSQL-style target times", () => {
    const active = resolver.resolve(
      [weekly("12:00:45")],
      new Date("2026-07-27T15:50:00Z"),
    );

    expect(active).toMatchObject({
      targetAt: "2026-07-27T16:00:45.000Z",
      value: "10m 45s",
    });
  });

  it("rejects trailing text instead of treating it as a valid time", () => {
    expect(
      resolver.resolve(
        [weekly("12:00oops")],
        new Date("2026-07-27T15:50:00Z"),
      ),
    ).toBeNull();
  });
});
